keps/173-headless-service-per-group/README.md

```diff
@@ -0,0 +1,297 @@
+      - [Integration tests](#integration-tests)
+      - [e2e tests](#e2e-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+LeaderWorkerSet creates a single headless service for the entire object.
+This KEP is about allowing one to create a headless service per each replica.
+
+## Motivation
+
+The number of pods per service is finite. At large scale, it is possible to exhaust this limit.
+See [DNS AT Scale](https://gist.github.com/aojea/32aeaa86aacebcdd93596ecb70fcba4f) for a more indepth explanation.
+
+With large number of replicas and size, we should allow one to create a service per replica.
+
+### Goals
+
+- Each group of leader-worker replicas can have a dedicated service
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- Providing an API to configure service per replicas.
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a user running a large scale LWS job, I want to have each leader-worker group with its own service.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### API
+
+```golang
+type LeaderWorkerSetSpec struct {
```

Comment:
should we allow more configurable API here? 
type ServiceConfiguration struct {
   Type ServiceConfigurationType
}
type ServiceConfigurationType string
const (
 DefaultHeadlessService ServiceConfigurationType = "DefaultHeadlessService"
 HeadlessServicePerReplica ServiceConfigurationType = "ServiceConfigurationType"
)

Reply:
Not sure. I sorta took the approach that if there isn’t a concrete use case for this, we shouldn’t support it yet. YAGNI.

Reply:
Two suggestions: 
1) To ensure that the API is future proof, I think we should encapsulate this in a `NetworkConfig` struct. We do the same in JobSet.
2) Since the headless service is an implementation detail of how we create dns host names of the pods, I think exposing that in the API is probably not the best approach. Instead, we can have the parameter represent the dns subdomain that will be created, shared vs unique per replica, and the consequence of that is shared vs one-per-replica headless service.

```
type NetworkConfig struct {
  SubdomainPolicy SubdomainPolicy
}
type SubdomainPolicy string
const (
  SubdomainShared SubdomainPolicy = "Shared"
  SubdomainUniquePerReplica SubdomainPolicy = "UniquePerReplica"
)
```

Reply:
Are you saying that you don't really want to have a single headless service anymore in general?

Reply:
Yes, this is generally a more scalable approach.

Reply:
What would you suggest for a transition strategy?

This would be a breaking change I think. 

We set environment variables on Pods to match the single headless service. I think that new lws on upgrades would need to change to the new service.

Reply:
> We set environment variables on Pods to match the single headless service.

I don't think this is how it works. Were do we do that?

Reply:
https://github.com/kubernetes-sigs/lws/blob/main/pkg/utils/pod/pod_utils.go#L115

Reply:
ok, so it turns out that the service name is actually the subdomain, we can't set a subdomain that is different from the headless service name :(

because of this, the API suggestion above in https://github.com/kubernetes-sigs/lws/pull/177#discussion_r1696170639 is I think what we can go with

Reply:
Do you want to default to UniquePerReplica from 0.4.0 on?

Reply:
Unfortunately we can't, the default must be what we have right now.

We need to think of whether we want to make this an immutable field, and if so how we will handle such an update.

Reply:
Can you please update the API as suggested in https://github.com/kubernetes-sigs/lws/pull/177/files#r1696170639? I think it better convoy what the user facing impact will be when it is used.